File: backend/ml/classifier/handwritten/nn.py
import random as rand

from functions import *

"""
A neuron (or a node) in the neural network. Each neuron has a state; a set of 
values, i.e. input, output, and corresponding derivatives (deltas), that change
after each round of forward and backward propagation.
"""
class Neuron:
    bias = 0.1
    activation = None


    inputs = []
    outputs = []
    inputVal = None
    outputVal = None

    # See `backprop()`; we use the distributive property of derivatives
    inputDelta = 0
    outputDelta = 0
    accDelta = 0
    numAccumulatedDelta = 0

    def __init__(self, id, activation, initZero=False):
        self.id = id
        if (initZero): self.bias = 0

    def update(self):
        self.inputVal = self.bias
        for i in self.inputs:
            self.inputVal += i.weight * i.source.outputVal
        self.outputVal = self.activation.out(self.inputVal)

"""
A wire (or link) in the neural network. Each wire has a weight, a source neuron
and a destination neuron. It also, similar to a neuron, has an internal state;
an error derivative w.r.t. a particular input that gets updated after each run
of backward propagation.
"""
class Wire:
    weight = rand.randrange(-0.5, 0.5)

    errorDelta = 0
    accErrorDelta = 0
    numAccumulatedDelta = 0

    regularisation = None
    
    def __init__(self, source, dest, regularisation, initZero=False):
        self.id = source.id + "->" + dest.id
        self.source = source
        self.dest = dest
        if (initZero): self.weight = 0

"""
Builds a neural network.

 :param networkShape:       The shape of the network. E.g. [1, 2, 3, 1] means
                            the network will have one input neuron, 2 neurons
                            in first hidden layer, 3 neurons in second hidden
                            layer and 1 output neuron.
 :param activation:         The activation function of every hidden neuron.
 :param outputActivation:   The activation function for the output neurons.
 :param regularization:     The regularisation function that computes a penalty
                            for a given weight (parameter) in the network. If
                            null, there will be no regularization.
 :param inputIds:           List of ids for the input neurons.
 :return:                   The network, as a 2D array of neurons
"""
def buildNetwork(networkShape, activation, outputActivation, regularisation, 
                 inputIds, initZero=False):
    id = 1
    network = []

    # Build the input layer
    currLayer = []
    for i in range(0, networkShape[0]):
        currLayer.append(Neuron(inputIds[i], activation, initZero))
    network.append(currLayer)

    # Build the hidden layers
    for i in range(1, len(networkShape)-1):
        currLayer = []
        for j in range(0, networkShape[i]):
            currNeuron = Neuron(str(id), activation, initZero); id += 1
            # Build the wires from the previous layer to this layer
            for k in range(0, networkShape[i-1]):
                prevNeuron = network[i-1][k]; currWire = Wire(prevNeuron, currNeuron, regularisation, initZero)
                currNeuron.inputs.append(currWire); prevNeuron.outputs.append(currWire)
            currLayer.append(currNeuron)
        network.append(currLayer)

    # Build the output layer
    currLayer = []
    for i in range(0, networkShape[-1]):
        currNeuron = Neuron(str(id), activation, initZero); id += 1
        # Build finally the last set of wires
        for j in range(0, networkShape[-2]):
            prevNeuron = network[i-1][k]; currWire = Wire(prevNeuron, currNeuron, regularisation, initZero)
            currNeuron.inputs.append(currWire); prevNeuron.outputs.append(currWire)
        currLayer.append(currNeuron)
    network.append(currLayer)

    return network

"""
Runs a forward propagation of a given input through a given network.

 :param network:    The neural network
 :param inputs:     The input array (its size should match that of the network)
 :return:           The final output array of the network

This function modifies the internal state of the network.
"""
def forwardProp(network, inputs):
    inputLayer = network[0]
    if len(inputs) != len(inputLayer):
        raise Exception("Size incompatibility between network and inputs!")
    
    # Feed in the inputs
    for i in range(0, len(inputs)):
        inputNeuron = inputLayer[i]
        inputNeuron.outputVal = inputs[i]

    # Propagate the input
    for i in range(1, len(network)):
        currLayer = network[i]
        for j in range(0, len(currLayer)):
            currLayer[j].update()
    
    return network[-1]

"""
Runs a backward propagation of a given neural network.

 :param network:    A neural network
 :param label:      The set of expected values
 :param errorFn:    The error function used to compute loss

This function modifies the internal state of the network, namely, the error
derivatives w.r.t. each neuron and the weights of the wires.

Note.   The main idea here is to use `outputDelta` to sum up chain-rule paths to
        a neuron and distributively multiplty it with ∂a(n)/∂z(n) to store the
        full path in `inputDelta`
"""
def backProp(network, label, LossFn):

    # Find ∂C/∂a0
    outputLayer = network[-1]
    for i in range(0, outputLayer.size):
        outNeuron = outputLayer[i]
        outNeuron.outputDelta = LossFn.der(outNeuron.outputVal, label)
    
    # Iterate through each layer backwards
    for i in range(len(network), -1, -1):
        layer = network[i]

        """
        Find partial chain
           [[∂C/∂m] * [∂am/∂zm * ... * ∂a(n+1)/∂z(n+1) * ∂z(n+1)/∂an]] * [∂a(n)/∂z(n)]
        up to the total input z of a neuron in the nth layer.
        
        Note he idea is to perform this iteratively, noting that the sum up
        until ∂a(n+1)/∂z(n+1) can be substituted with n' = n+1.
        """
        for j in range(0, len(layer)):
            neuron = layer[j]
            neuron.inputDelta = neuron.outputDelta * neuron.activation.der(neuron.inputVal)
            neuron.accDelta += neuron.inputDelta
            neuron.numAccumulatedDelta += 1

        # Find the gradient of the weights of each wire
        for j in range(0, len(layer)):
            neuron = layer[j]
            for k in range(0, len(neuron.inputs)):
                wire = neuron.inputs[k]
                if (wire.isDead): continue
                wire.errorDelta = wire.source.outputVal * neuron.inputDelta
                wire.accErrorDelta += wire.errorDelta; wire.numAccumulatedDelta += 1
        
        if (i == 1): continue

        # Cumulate the chain-rule paths of influence to the next (backwards) layer
        prevLayer = network[i-1]
        for j in range(0, len(prevLayer)):
            prevNeuron = prevLayer[j]
            prevNeuron.outputDelta = 0
            for k in range(0, len(prevNeuron.outputs)):
                neuronLink = prevNeuron.outputs[k]
                prevNeuron.outputDelta += neuronLink.weight * neuronLink.dest.inputDelta


    return

"""
Updates the parameters of a given network given the previously accumulated
partial derivatives
"""
def updateParams(network, learningRate, regLambda):
    # Ignore the input layer
    for i in range(1, network.size):
        layer = network[i]
        
        
        for j in range(0, layer.size):
            # Step the bias
            neuron = layer[j]
            if (neuron.numAccumulatedDelta <= 0): continue
            neuron.bias -= learningRate * (neuron.accDelta / neuron.numAccumulatedDelta)
            neuron.accDelta = 0; neuron.numAccumulatedDelta = 0

            # Step the gradient of the wires feeding into this neuron
            for k in range(0, len(neuron.inputs)):
                wire = neuron.inputs[k]
                if (wire.isDead): continue
                if (wire.numAccumulatedDelta <= 0): continue 
                wire.weight -= learningRate * (wire.accDelta / wire.numAccumulatedDelta)

                # Process regularisation; derive the penalty component of the
                # cost w.r.t. the weight, i.e. ∂C/∂W = ∂(L+P)/∂W = ∂L/∂W + ∂P/∂W (=dP(W)) [P_W : W⊆R->R],
                # added to desensitise the parameters (lower weight between neurons) (fixing the 'ridge').
                regDer = wire.regularisation.der if (wire.regularisation is not None) else 0
                regWeight = wire.weight - (learningRate * regLambda * regDer)
                if (wire.regularisation == Regularisations.L1):
                    # For L1 regularisation, check if the desensitisation then
                    # crosses 0; if so, kill. (Useful for useless parameters).
                    if (wire.weight * regWeight < 0):
                        wire.weight = 0; wire.isDead = True
                else: # Otw for L2, 
                    wire.weight = regWeight
                
                wire.accDelta = 0; wire.numAccumulatedDelta = 0
    
    return

File: src/nn/nn.ts
import * as fns from "./functions";

export class Node {
  id: string;
  bias: number;
  activation: fns.Activation;
  totalInput: number;
  output: number;
  inputs: Link[];
  outputs: Link[];

  // Final derivatives; `derPath` should end up as `∂a/∂z` so that `∂z/∂w` is
  //                    performed in `updateParams`
  derA: number;
  derZ: number;
  accDer: number;
  numAccDer: number;


  constructor(id: string, activation: fns.Activation) {
    this.id = id;
    this.activation = activation;
    this.totalInput = 0;
    this.output = 0;
    this.inputs = [];
    this.outputs = [];
    this.bias = 0.1;

    this.derA = 0; this.derZ = 0; this.accDer = 0; this.numAccDer = 0;
  }

  initBias(bias: fns.Bias) {
    this.bias = bias.gen(this.inputs.length);;
  }

  /**
   * Note; thhis update algorithm for forward propagation is from
   * https://github.com/tensorflow/playground/blob/master/src/nn.ts.
   */
  updateOutput(): number {
    this.totalInput = this.bias;
    for (let j = 0; j < this.inputs.length; j++) {
      let link = this.inputs[j];
      this.totalInput += link.weight * link.source.output;
    }
    this.output = this.activation.output(this.totalInput);
    return this.output;
  }
}

export class Link {
  id: string;
  
  source: Node;
  dest: Node;
  weight: number;

  der: number;
  accDer: number;
  numAccDer: number;

  constructor(id: string, source: Node, dest: Node) {
    this.id = id;

    this.source = source;
    this.dest = dest;
    this.weight = Math.random() - 0.5;

    this.der = 0; this.accDer = 0; this.numAccDer = 0;
  }
}

export function buildNetwork(
  shape: number[],
  activation: fns.Activation,
  outputActivation: fns.Activation): Node[][]
{
  let network: Node[][] = [];
  let id = 1;

  // First build the input layer
  let inputLayer = [];
  for (let i = 0; i < shape[0]; i++) {
    const node = new Node(String(id), activation);
    node.initBias(fns.Bias.ZERO);
    id++;
    inputLayer.push(node);
  }
  network.push(inputLayer);
  
  // Next build the hidden layers and output layer
  for (let i = 1; i < shape.length; i++) {
    let layer = [];
    for (let j = 0; j < shape[i]; j++) {
      const node = new Node(String(id), i === shape.length - 1 ? outputActivation : activation);
      id++;
      // Links: add links between this node and each previous node
      for (let k = 0; k < shape[i - 1]; k++) {
        const prev = network[i - 1][k];
        const link = new Link(prev.id + " -> " + node.id, prev, node);
        node.inputs.push(link);
        prev.outputs.push(link);
      }
      node.initBias(fns.Bias.HE);
      layer.push(node);
    }
    network.push(layer);
  }

  return network;
}

export function forwardProp(network: Node[][], inputs: number[]) {
  let inputLayer = network[0];
  if (inputs.length != inputLayer.length) {
    throw new Error(`Size error: incompatibility between given input size ${inputs.length} and network input size ${inputLayer.length}!`);
  }

  // First: update the input layer
  for (let i = 0; i < inputs.length; i++) {
    const node = inputLayer[i];
    node.output = inputs[i];
  }
  
  for (let i = 1; i < network.length; i++) {
    const layer = network[i];
    for (let j = 0; j < layer.length; j++) {
      const node = layer[j];
      node.updateOutput();
    }
  }

  return network[network.length - 1];
}

export function backProp(network: Node[][], labels: number[], lossFn: fns.Loss) {
  const outputLayer = network[network.length - 1];
  if (labels.length != outputLayer.length) {
    throw new Error(`Size error: incompatibility between given input size ${labels.length} and network input size ${outputLayer.length}!`);
  }

  //    A. Compute ∂C/∂a(L), for each output neuron
  const numOutputs = outputLayer.length;
  for (let i = 0; i < numOutputs; i++) {
    const node = outputLayer[i];
    const label = labels[i];
    node.derA = lossFn.der(node.output, label, numOutputs);
  }

  /**   B.  Iterate through the layers backwards, and for each node of each layer...
   *          1)  Perform   ∂a(L)/∂z(L) = act'(z(L))      using the `derA` value stored to give `derZ`;
   *          2)  Perform   ∂z(L)/∂W(L_{i->j}) = a(L-1)   to each weight extending into that node.
   *          3)  Perform   Σ ( ∂z(L)/∂a(L-1) ) = W       over each node in the current layer,
   *              and store in the next (backward) layer's nodes' `derA`.
   *    Reset temporary `derPath` value backpropagation before each assignment.
  */
  for (let i = network.length - 1; i > 0; i--) {
    
    const layer = network[i];
    for (let j = 0; j < layer.length; j++) {
      const node = layer[j];
      node.derZ = node.derA * node.activation.der(node.totalInput);
      node.accDer += node.derZ;
      node.numAccDer++;
    }

    for (let j = 0; j < layer.length; j++) {
      const node = layer[j];
      for (let k = 0; k < node.inputs.length; k++) {
        const link = node.inputs[k];
        link.der = node.derZ * link.source.output;
        link.accDer += link.der;
        link.numAccDer++;
      }
    }

    if (i === 1) {
      continue;
    }

    const prevLayer = network[i - 1];
    for (let j = 0; j < prevLayer.length; j++) {
      const prevNode = prevLayer[j];
      prevNode.derA = 0;
      for (let k = 0; k < prevNode.outputs.length; k++) {
        const link = prevNode.outputs[k];
        prevNode.derA += link.dest.derZ * link.weight;
      }
    }
  }
}

export function updateParams(network: Node[][], learningRate: number) {
  for (let i = 1; i < network.length; i++) {
    const layer = network[i];
    for (let j = 0; j < layer.length; j++) {
      const node = layer[j];
      if (node.numAccDer > 0) {
        node.bias -= learningRate * (node.accDer / node.numAccDer);
        node.accDer = 0;
        node.numAccDer = 0;
      }
      for (let k = 0; k < node.inputs.length; k++) {
        const link = node.inputs[k];
        if (link.numAccDer > 0) {
          link.weight -= learningRate * (link.accDer / link.numAccDer);
          link.accDer = 0; 
          link.numAccDer = 0;
        }
      }
    }
  }
}

/**
 * CONVOLUTION NEURAL NETWORK
 */
class Filter {
  height: number;
  width: number;
  bias: number;
  stride: number;
  activation: fns.Activation;
  kernel: number[][];

  constructor(height: number, width: number, bias: number, stride: number, activation: fns.Activation) {
    this.height = height;
    this.width = width;
    this.bias = bias;
    this.stride = stride;
    this.activation = activation;
    this.kernel = init2DArray(height, width);
  }
}

class Pool {
  height: number;
  width: number;
  stride: number;

  constructor(height: number, width: number, stride: number) {
    this.height = height;
    this.width = width;
    this.stride = stride;
  }
}

class convolutionalLayer {
  filter: Filter;
  pool: Pool;
  featureMap: number[][];
  
  constructor(filter: Filter, pool: Pool) {
    this.filter = filter;
    this.pool = pool;
    this.featureMap = [];
  }

  convolute(image: number[][]): void {
    let featureMap = [];
    for (let i = 0; i < image.length - this.filter.height; i++) {
      let row = image[i];
      let resRow = [];
      for (let j = 0; j < row.length - this.filter.width; j += this.filter.stride) {
        let sum = 0;
        for (let k = i; k < this.filter.height; k++) {
          for (let l = j; l < this.filter.width; l++) {
            sum += image[k][l] * this.filter.kernel[k][l];
          }
        }
        sum += this.filter.bias;
        sum = this.filter.activation.output(sum);
        resRow.push(sum);
      }
      featureMap.push(resRow);
    }
    this.featureMap = featureMap;
  }

  maxPool(): number[][] {
    let inputs = [];
    for (let i = 0; i < this.featureMap.length - this.pool.height; i += this.pool.height) {
      let row = this.featureMap[i];
      let resRow = [];
      for (let j = 0; j < row.length - this.pool.width; j += Math.max(this.pool.width, this.pool.stride)) {
        let max = -Infinity;
        for (let k = i; i < this.pool.height; i++) {
          for (let l = j; l < this.pool.width; j++) {
            let num = this.featureMap[k][l];
            if (num > max) {
              max = num;
            }
          }
        }
        resRow.push(max);
      }
      inputs.push(resRow);
    }
    return inputs;
  }
}

export function cnnForwardProp(image: number[][], convNetwork: convolutionalLayer[], network: Node[][]) {
  let forwardImage = image;
  for (let i = 0; i < convNetwork.length; i++) {
    const layer = convNetwork[i];
    layer.convolute(forwardImage);
    layer.maxPool();
    forwardImage = layer.featureMap;
  }
  return forwardProp(network, flat(forwardImage));
}

export function cnnBackProp(network: Node[][], target: number[], lossFn: fns.Loss) {
  backProp(network, target, lossFn);
}

/**
 * HELPER FUNCTIONS
 */
function flat<Type>(arr: Type[][]): Type[] {
  let flatArr = arr.reduce((acc, curr) => acc.concat(curr), []);
  return flatArr;
}

function init2DArray(height: number, width: number): number[][] {
  const result: number[][] = [];
  for (let i = 0; i < height; i++) {
    const row = [];
    for (let j = 0; j < width; j++) {
      const val = Math.random() - 0.5;
      row.push(val);
    }
    result.push(row);
  }
  return result;
}